src/setup-beam.js

```diff
@@ -333,62 +342,63 @@ async function getRebar3Versions() {
     'https://api.github.com/repos/erlang/rebar3/releases?per_page=100',
     [1, 2, 3],
   )
-  const rebar3VersionsListing = []
+  const rebar3VersionsListing = new Map()
   resultJSONs.forEach((resultJSON) => {
     jsonParseAsList(resultJSON)
       .map((x) => x.tag_name)
-      .sort()
-      .forEach((v) => rebar3VersionsListing.push(v))
+      .forEach((ver) => rebar3VersionsListing.set(ver, ver))
   })
+
   return rebar3VersionsListing
 }
 
 function isStrictVersion() {
   return getInput('version-type', false) === 'strict'
 }
 
-function getVersionFromSpec(spec, versions, maybePrependWithV0) {
-  let version = null
-
-  if (spec.match(/rc/) || isStrictVersion()) {
-    version = spec
-  }
-
-  if (version === null) {
-    // We keep a map of semver => "spec" in order to use semver ranges to find appropriate versions
-    const versionsMap = versions.sort(sortVersions).reduce((acc, v) => {
-      if (!v.match(/rc/)) {
-        // release candidates are opt-in
-        acc[maybeCoerced(v)] = v
-      }
-      return acc
-    }, {})
-    const rangeForMax = semver.validRange(spec)
-    if (rangeForMax) {
-      version =
-        versionsMap[semver.maxSatisfying(Object.keys(versionsMap), rangeForMax)]
+function getVersionFromSpec(spec0, versions0) {
+  const versions = Array.from(versions0.keys())
+  versions.forEach((version, index, array) => {
+    let manipulatedVersion
+    if (isStrictVersion() || isRC(version)) {
+      manipulatedVersion = maybeRemoveVPrefix(version)
     } else {
-      version = versionsMap[maybeCoerced(spec)]
+      manipulatedVersion = maybeCoerced(version)
     }
-  }
+    array.splice(index, 1, manipulatedVersion)
+  })
+  versions.sort(sortVersions)
 
-  let v = version === null || version === undefined ? null : version
-  if (maybePrependWithV0 && v != null) {
-    v = maybePrependWithV(v)
-  }
+  let version = null
+  const isStrictVersionOrRC = isStrictVersion() || isRC(spec0)
+  if (isStrictVersionOrRC) {
+    const spec = maybeRemoveVPrefix(spec0)
+    if (versions.includes(spec)) {
+      version = spec
+    }
+  } else {
+    const rangeForMax = semver.validRange(spec0)
 
-  if (!versions.includes(v)) {
-    v = null
+    if (rangeForMax) {
+      version = semver.maxSatisfying(versions, rangeForMax)
+    } else if (semver.validRange(version) === null) {
+      throw new Error(
+        `you have to set version-type=strict when using a semver-invalid version (got ${spec0})`,
+      )
+    }
   }
 
-  return v
+  return version || null
 }
 
 function maybeCoerced(v) {
```

Comment:
Most of the big changes are in this function... and `getVersionFromSpec/2` above.